### DISCUSSION TITLE
Oracledb circular import problem

### DISCUSSION THREAD

**Discussion by FantasticMrFox630** [oracle/python-oracledb#154] [Q&A]
I've freshly installed python-oracledb and "import oracledb" is giving me a circular import error. 

oracledb\__init___.py contains "from .connection import connect, Connection"

oracledb\connection.py contains "import oracledb"

Am I missing something here?

**Comment by cjbj** [comment]
Do you have your own application file called oracledb.py?

**Reply by anthony-tuininga** [reply]
Can you supply the full traceback that you are getting?

**Reply by FantasticMrFox630** [reply]
This is the traceback:

```
Traceback (most recent call last):

  File "...", line 8, in <module>
    import oracledb

  File "C:\Anaconda3_2021.11\lib\site-packages\oracledb\__init__.py", line 38, in <module>
    from .connection import connect, Connection

  File "C:\Anaconda3_2021.11\lib\site-packages\oracledb\connection.py", line 45, in <module>
    from . import base_impl, thick_impl, thin_impl

ImportError: cannot import name 'base_impl' from partially initialized module 'oracledb' (most likely due to a circular import) (C:\Anaconda3_2021.11\lib\site-packages\oracledb\__init__.py)
```

**Reply by anthony-tuininga** [reply]
Which version? Which Python version? Do you get the same problem without using Anaconda?

**Reply by FantasticMrFox630** [reply]
oracledb version 1.2.2. I tried with both Python 3.11 and Python 3.9 (Anaconda).

**Reply by anthony-tuininga** [reply]
Can you share the contents of your installation directory? And was Python 3.11 tried *without* Anaconda? It is possible that Anaconda is modifying the installation in some fashion.

**Reply by anthony-tuininga** [reply]
One other possibility: the oracledb module you imported is 32-bit but the Python you are using is 64-bit (or vice versa). The inability to load the binary submodule (base_impl) can result in the error you are getting. I'll have to see if there is a way to get a more useful error message for this situation!

**Reply by FantasticMrFox630** [reply]
I am using 64-bit Python. Not sure how to check that on the oracledb module. 

Python 3.11 was tried without Anaconda and 3.9 with Anaconda. Here are the contents of my installation directory:

```
Directory of C:\Anaconda3_2021.11\Lib\site-packages\oracledb

03/03/2023  08:36 AM    <DIR>          .
03/03/2023  08:36 AM    <DIR>          ..
03/03/2023  08:36 AM            19,011 aq.py
03/03/2023  08:36 AM           743,936 base_impl.cp311-win_amd64.pyd
03/03/2023  08:36 AM            49,404 connection.py
03/03/2023  08:36 AM            30,713 connect_params.py
03/03/2023  08:36 AM             4,070 constants.py
03/03/2023  08:36 AM             2,909 constructors.py
03/03/2023  08:36 AM            35,250 cursor.py
03/03/2023  08:36 AM            11,854 dbobject.py
03/03/2023  08:36 AM             1,827 defaults.py
03/03/2023  08:36 AM             5,403 driver_mode.py
03/03/2023  08:36 AM             3,130 dsn.py
03/03/2023  08:36 AM            20,629 errors.py
03/03/2023  08:36 AM             1,811 exceptions.py
03/03/2023  08:36 AM             6,076 lob.py
03/03/2023  08:36 AM            35,229 pool.py
03/03/2023  08:36 AM            31,290 pool_params.py
03/03/2023  08:36 AM                 0 py.typed
03/03/2023  08:36 AM            27,497 soda.py
03/03/2023  08:36 AM            11,078 subscr.py
03/03/2023  08:36 AM           614,400 thick_impl.cp311-win_amd64.pyd
03/03/2023  08:36 AM           967,168 thin_impl.cp311-win_amd64.pyd
03/03/2023  08:36 AM             3,382 utils.py
03/03/2023  08:36 AM             6,394 var.py
03/03/2023  08:36 AM             1,533 version.py
03/03/2023  08:36 AM             3,881 __init__.py
03/03/2023  08:36 AM    <DIR>          __pycache__
              25 File(s)      2,637,875 bytes
```

**Reply by Karthikeyan-GR** [reply]
@anthony-tuininga 
Facing same problem. Using python version 3.12 and oracledb version 2.0.1.
Directory structure looks fine. When I execute my function in local, It works good with no error. Whereas if I zip the same directory and uploads it  in AWS lambda I am facing circular import error while executing the function. I am not facing this issues with any other libraries. Kindly let me know the ways to work it out.

**Reply by anthony-tuininga** [reply]
I don't know if I can help you out as I don't use AWS lambda. Hopefully someone else with that knowledge can help you out. As noted earlier in this issue you will need to find out *why* the submodule is not being loaded. It would be useful to know the full traceback, however. I might be able to suggest some changes to make to the driver to help you diagnose the issue.

**Reply by kishorkumarg** [reply]
@Karthikeyan-GR , is this issue resolved, I am facing the same issue, I am trying to resolve since two days, it's really urgent, can you please help me out.

Thank you in advance

**Reply by Karthikeyan-GR** [reply]
@kishorkumarg  In my case, I was building python with the package file in windows. So with Lambda, it thrown circular import error.
Later understood that Lambda is based on linux arch, so I built the complete package file in linux and used in Lambdas, it worked.

**Reply by anthony-tuininga** [reply]
I wonder if we can get a better traceback by just changing the order of the imports in `oracledb/__init__.py` -- specifically so that the import of the base_impl, thin_impl and thick_impl submodules occurs first? Can someone who is having this problem try that and let me know if that helps?

**Reply by DenisTru** [reply]
> I wonder if we can get a better traceback by just changing the order of the imports in `oracledb/__init__.py` -- specifically so that the import of the base_impl, thin_impl and thick_impl submodules occurs first? Can someone who is having this problem try that and let me know if that helps?


I moved these imports to the top and no change 
from . import base_impl, thick_impl, thin_impl
from .base_impl import (...)
from .thick_impl import (..)

fresh ec2 instance RHEL 9 trying to run off embedded python(3.8.19) in LibreOffice

  File "/home/ec2-user/test_flask_app2/webserver.py", line 9, in <module>
    import oracledb
  File "/home/ec2-user/.local/lib/python3.8/site-packages/oracledb/__init__.py", line 30, in <module>
    from . import base_impl, thick_impl, thin_impl
ImportError: cannot import name 'base_impl' from partially initialized module 'oracledb' (most likely due to a circular import) (/home/ec2-user/.local/lib/python3.8/site-packages/oracledb/__init__.py)

**Reply by anthony-tuininga** [reply]
I just pushed some changes to the code to eliminate all circular imports. Can you build from source and try again? And let me know whether it is successful or not and we discover the *real* error! Thanks.

**Comment by cjbj** [comment]
Changes that eliminate circular imports landed in 2.3.0.

**Comment by AkshatJain1402** [comment]
hey, 
I am also facing the same issue of circular imports 

Error Category: IMPORT_ERROR; Failed Line Number: 8; ImportError: cannot import name 'base_impl' from partially initialized module 'oracledb' (most likely due to a circular import) (/tmp/glue-job-13151463220786226397/extra-py-files/oracledb-2.5.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_28_x86_64.whl/oracledb/__init__.py)


we are trying to use the oracledb module in glue job , downloaded it's linux compatible wheel files as suggested in aws documentations as well as on internet 
also tried using other pip downloads of oracle db that were directory with following structure 
same error occured ... 

IF you can provide or suggest any solution it would be really helpful



03/03/2023  08:36 AM            19,011 aq.py
03/03/2023  08:36 AM           743,936 base_impl.cp311-win_amd64.pyd
03/03/2023  08:36 AM            49,404 connection.py
03/03/2023  08:36 AM            30,713 connect_params.py
03/03/2023  08:36 AM             4,070 constants.py
03/03/2023  08:36 AM             2,909 constructors.py
03/03/2023  08:36 AM            35,250 cursor.py
03/03/2023  08:36 AM            11,854 dbobject.py
03/03/2023  08:36 AM             1,827 defaults.py
03/03/2023  08:36 AM             5,403 driver_mode.py
03/03/2023  08:36 AM             3,130 dsn.py
03/03/2023  08:36 AM            20,629 errors.py
03/03/2023  08:36 AM             1,811 exceptions.py
03/03/2023  08:36 AM             6,076 lob.py
03/03/2023  08:36 AM            35,229 pool.py
03/03/2023  08:36 AM            31,290 pool_params.py
03/03/2023  08:36 AM                 0 py.typed
03/03/2023  08:36 AM            27,497 soda.py
03/03/2023  08:36 AM            11,078 subscr.py
03/03/2023  08:36 AM           614,400 thick_impl.cp311-win_amd64.pyd
03/03/2023  08:36 AM           967,168 thin_impl.cp311-win_amd64.pyd
03/03/2023  08:36 AM             3,382 utils.py
03/03/2023  08:36 AM             6,394 var.py
03/03/2023  08:36 AM             1,533 version.py
03/03/2023  08:36 AM             3,881 __init__.py
03/03/2023  08:36 AM    <DIR>          __pycache__

**Reply by cjbj** [reply]
The above information showed you used both Linux and Windows variants of the driver.

What output does this produce when you run it in Glue?

```
import sys
import platform

print("platform.platform:", platform.platform())
print("sys.maxsize > 2**32:", sys.maxsize > 2**32)
print("platform.python_version:", platform.python_version())
```
